.github/workflows/rust.yml

```diff
@@ -51,6 +53,28 @@ jobs:
     - run: rustup update
     - run: rustup target add ${{ matrix.sys.target }}
     - run: cargo install --locked --version 0.5.14 cargo-hack
-    - run: cargo hack --feature-powerset check --target ${{ matrix.sys.target }}
+    - run: cargo hack --feature-powerset check --locked --target ${{ matrix.sys.target }}
     - if: matrix.sys.test
-      run: cargo hack --feature-powerset test --target ${{ matrix.sys.target }}
+      run: cargo hack --feature-powerset test --locked --target ${{ matrix.sys.target }}
+
+  cargo-publish:
+    if: github.event_name == 'release'
+    needs: [complete]
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - run: rustup update
+    - run: cargo publish --locked
+      env:
+      - CARGO_REGISTRY_TOKEN: ${{ secrets.CARGO_REGISTRY_TOKEN }}
+
+  cargo-publish-dry-run:
+    if: github.event_name == 'push' && startsWith(github.ref_name, 'release-dry-run')
+    needs: [complete]
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - run: rustup update
+    - run: cargo publish --locked --dry-run
+      env:
+      - CARGO_REGISTRY_TOKEN: ${{ secrets.CARGO_REGISTRY_TOKEN }}
```

Comment:
Does the dry run need the token? It might be worth trying without it just to reduce the places that get the token to the absolute minimal. But if it's needed lgtm.